SI-5919 TypeTags and Exprs should be serializable

phaller opened the pull request:
- Make TypeCreator and TreeCreator extend Serializable.
- When replacing a SerializedTypeTag with a TypeTag or WeakTypeTag,
  do not use scala.reflect.runtime.universe.rootMirror, since
  it is unlikely to find user classes; instead, create a runtime
  mirror using the context ClassLoader of the current thread.
  Use the same logic for SerializedExpr.
- Remove writeObject/readObject methods from SerializedTypeTag
  and SerializedExpr since they are unused.
- Add @throws annotation on writeReplace and readResolve methods.
- Handle SecurityException if the current thread cannot access the
  context ClassLoader.
- To make type tags of primitive value classes serializable, make
  PredefTypeCreator a top-level class. Otherwise, it would
  retain a reference to the enclosing Universe,
  rendering the TypeCreator non-serializable.

xeno-by commented on src/reflect/scala/reflect/api/TreeCreator.scala:
```diff
@@ -2,12 +2,12 @@ package scala
+ * @group ReflectionAPI
  */
-abstract class TreeCreator {
+abstract class TreeCreator extends Serializable {
```
How's it going to work for more complex tree and type creators, e.g. those that create auxiliary symbols?

phaller replied:
See `writeReplace` in `TypeTagImpl`, `WeakTypeTagImpl`, and `PredefTypeTag`. The instances that are actually serialized are instances of `SerializedTypeTag`. Tree and type creators are serializable also for more complex trees/types.

xeno-by commented on src/reflect/scala/reflect/api/TypeTags.scala:
```diff
@@ -341,20 +335,23 @@ trait TypeTags { self: Universe =>
+  import scala.reflect.runtime.universe.{TypeTag, WeakTypeTag, runtimeMirror}
+  @throws(classOf[ObjectStreamException]) private def readResolve(): AnyRef = {
+    val loader: ClassLoader = try {
+      Thread.currentThread().getContextClassLoader()
```
Why's this necessary?

phaller replied:
To be able to refer to types outside the Scala library. I'll add a test case which demonstrates this.

xeno-by replied:
What if someone can't set a context classloader?

phaller replied:
Then the previous behavior kicks in, meaning the rootMirror will be used.

xeno-by replied:
I mean what if someone wants to customize the behavior, but can't set the context classloader? What could be other ways to solve the problem?

phaller commented:
FYI @heathermiller

phaller commented:
PLS REBUILD/pr-scala@e9d0507

scala-jenkins commented:
(kitty-note-to-self: ignore 45337374)
:cat: Roger! Rebuilding pr-scala for e9d0507d. :rotating_light:

phaller commented:
PLS REBUILD/pr-scala@55a4b3b

scala-jenkins commented:
(kitty-note-to-self: ignore 45346589)
:cat: Roger! Rebuilding pr-scala for 55a4b3ba. :rotating_light:

phaller commented:
PLS REBUILD/pr-scala@8a19c27

scala-jenkins commented:
(kitty-note-to-self: ignore 47883296)
:cat: Roger! Rebuilding pr-scala for 8a19c275. :rotating_light:

phaller commented:
Ping @gkossakowski 
It would be great if you could have a look at it in the next few days, since this is supposed to be part of the 2.11.2 release (5 days left for merging changes). Thanks a lot!

retronym commented:
This PR doesn't merge cleanly.

phaller commented:
OK, will rebase.

phaller committed:
SI-5919 TypeTags and Exprs should be serializable

- Make TypeCreator and TreeCreator extend Serializable.
- When replacing a SerializedTypeTag with a TypeTag or WeakTypeTag,
  do not use scala.reflect.runtime.universe.rootMirror, since
  it is unlikely to find user classes; instead, create a runtime
  mirror using the context ClassLoader of the current thread.
  Use the same logic for SerializedExpr.
- Remove writeObject/readObject methods from SerializedTypeTag
  and SerializedExpr since they are unused.
- Add @throws annotation on writeReplace and readResolve methods.
- Handle SecurityException if the current thread cannot access the
  context ClassLoader.
- To make type tags of primitive value classes serializable, make
  PredefTypeCreator a top-level class. Otherwise, it would
  retain a reference to the enclosing Universe,
  rendering the TypeCreator non-serializable.

Binary compatibility:

- Keep nested PredefTypeCreator class to avoid backward binary
  incompatible change.
- Keep `var` modifiers on the class parameters of
  SerializedTypeTag for backward binary compatibility.
- Adds filter rules to forward binary compatibility whitelist:
  - `TypeCreator`, `PredefTypeCreator`, and `TreeCreator` must now
    extend from `Serializable`.
  - Must have new class `scala.reflect.api.PredefTypeCreator`
    to avoid problematic outer reference.

xeno-by commented on bincompat-forward.whitelist.conf:
```diff
@@ -234,6 +234,23 @@ filter {
+    {
+        matchName="scala.reflect.api.PredefTypeCreator"
+        problemName=MissingClassProblem
     }
```
So this is going to break bincompat for scala-reflect.jar?

phaller replied:
These entries are only necessary, because the listed classes now extend `Serializable`. Thus, it does not break bincompat for `scala-reflect.jar` as far as I can see.

phaller replied:
According to [JLS Chapter 13, section 13.4.4](http://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html#jls-13.4.4), extending from `Serializable` does not break bincompat for pre-existing binaries. (Reason: the total set of superclasses or superinterfaces, respectively, of the class type loses no members.)

xeno-by commented on bincompat-forward.whitelist.conf:
```diff
@@ -234,6 +234,23 @@ filter {
+    {
+        matchName="scala.reflect.api.TypeCreator"
+        problemName=MissingTypesProblem
+    },
```
Why are TreeCreator and TypeCreator reported missing?

phaller replied:
These classes themselves are obviously not missing. The whitelist talks only about forward bincompat; compiling a class `A` against a class, say, `TreeCreator` that in version x+1 has a new superclass with additional members can cause a linkage error when `A` is used with version x of `TreeCreator`. Therefore, forward bincompat is an issue in this case. Now, in our case the new superinterface `Serializable` does not add new members, and thus no such forward bincompat issue can arise. Therefore, the reported problems are due to the fact that MIMA is not precise in this case.

phaller commented:
Ping @gkossakowski @retronym

gkossakowski commented:
Hey!

I'm offline until Sunday. I'll review it next week if nobody beats me first.

W dniu czwartek, 10 lipca 2014 Philipp Haller notifications@github.com
napisał(a):

> Ping @gkossakowski https://github.com/gkossakowski @retronym
> https://github.com/retronym
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/scala/scala/pull/3817#issuecomment-48591452.

## 

Grzegorz Kossakowski
Scalac hacker at Typesafe http://www.typesafe.com/
twitter: @gkossakowski http://twitter.com/gkossakowski
github: @gkossakowski http://github.com/gkossakowski

phaller commented:
Thanks for the heads-up, @gkossakowski.

@lrytz: does the time until 2.11.2 permit a review next week, or shall we try to find someone else to review it?

lrytz commented:
> @lrytz: does the time until 2.11.2 permit a review next week, or shall we try to find someone else to review it?

Let's discuss tomorrow in Lausanne. The goal is to have everything merged by Monday evening.

lrytz commented on src/reflect/scala/reflect/api/TypeTags.scala:
```diff
@@ -341,20 +344,27 @@ trait TypeTags { self: Universe =>
 }
 
+// This class should be final, but we can't do that in Scala 2.11.x without breaking
+// binary incompatibility.
```
can we make it `@deprecatedInheritance` in 2.11, or do we have to wait until 2.12?

phaller replied:
`SerializedTypeTag` is `private[scala]`, so the deprecation warning would not actually help clients of the reflection library. Maybe the easiest would be to simply make the class final in 2.12?

lrytz replied:
True, agreed.

lrytz commented on src/reflect/scala/reflect/api/TypeTags.scala:
```diff
@@ -341,20 +344,27 @@ trait TypeTags { self: Universe =>
+// This class should be final, but we can't do that in Scala 2.11.x without breaking
+// binary incompatibility.
+// Since instances of this class are serialized, this class should have a
+// SerialVersionUID annotation.
```
Should that be done in 2.12? Could you log an issue and later create a follow-up PR against 2.12.x?

phaller replied:
Created an issue: https://issues.scala-lang.org/browse/SI-8722

phaller replied:
Actually, it might be better to add the `SerialVersionUID` annotation already now, in 2.11.2. The reason is that we will then be able to provide serialization stability throughout the 2.11.x series. And since type tags have not been serializable until this PR, we won't break serialization for existing code. What do you think?

lrytz replied:
Makes sense. The generated static `serialVersionUID` field is not directly accessible from Scala, so it is also backwards binary compatible. Maybe mima will flag it, not sure.

phaller committed:
Add SerialVersionUID to SerializedTypeTag and SerializedExpr

The reason for adding the SerialVersionUID annotations is to
be able to provide serialization stability throughout the 2.11.x
series. And since type tags (and exprs) have not been serializable
before, this does not break serialization for existing code.

phaller commented:
PLS REBUILD/pr-scala@6a360c4

scala-jenkins commented:
(kitty-note-to-self: ignore 48999724)
:cat: Roger! Rebuilding pr-scala for 6a360c43. :rotating_light:

gkossakowski commented on src/reflect/scala/reflect/api/Exprs.scala:
```diff
@@ -157,23 +158,21 @@ trait Exprs { self: Universe =>
       |import `scala.tools.reflect.Eval` and call `<your expr>.eval` instead.""".trim.stripMargin)
 
-    private def writeReplace(): AnyRef = new SerializedExpr(treec, implicitly[WeakTypeTag[T]].in(ru.rootMirror))
+    @throws(classOf[ObjectStreamException]) private def writeReplace(): AnyRef = new SerializedExpr(treec, implicitly[WeakTypeTag[T]].in(ru.rootMirror))
```
Normally we put annotations in separate line. Any reason to break out from that rule?

phaller replied:
No reason; I just fixed it.

gkossakowski commented on src/reflect/scala/reflect/api/Exprs.scala:
```diff
@@ -157,23 +158,21 @@ trait Exprs { self: Universe =>
+    } catch {
+      case se: SecurityException => null
+    }
+    val m = runtimeMirror(loader)
```
Does runtime mirror handle a null loader correctly?

phaller replied:
Yes, it does.

gkossakowski commented:
Could we deprecate setters for vars of `SerializedTypeTag`?

BTW. Excellent commit messages!

phaller commented:
Similar to the [comment about `@deprecatedInheritance`](https://github.com/scala/scala/pull/3817#discussion_r14880981) above, since `SerializedTypeTag` is `private[scala]` the deprecation would not help clients. Instead, I'd suggest to simply remove the setter in 2.12 (analogous to the decision for making the class final in 2.12).

retronym commented:
LGTM.

gkossakowski commented:
Agreed.

LGTM.

phaller commented:
Thanks!

gkossakowski commented:
PS. I leave pushing the merge button to @lrytz.
